fem/tevaluator.hpp

```diff
@@ -1417,6 +1417,3255 @@ class FieldEvaluator
+              const dof_layout_t &dof_layout, dof_data_t &dof_data) const
+   {
+      CalcT<0,Add>(qpt_layout.ind2(0),qpt_data,
+                   dof_layout, dof_data,
+                   Bt_1d_o.layout, Bt_1d_o,
+                   B_1d_c.layout,  B_1d_c);
+      CalcT<1,Add>(qpt_layout.ind2(1),qpt_data,
+                   dof_layout, dof_data,
+                   Bt_1d_c.layout, Bt_1d_c,
+                   B_1d_o.layout,  B_1d_o);
+      AddOrientation(dof_data);
+   }
+   // Multi-component gradient evaluation from DOFs to quadrature points.
+   // dof_layout is (TDOF x NumComp) and grad_layout is (TNIP x DIM x NumComp).
+   template <typename dof_layout_t, typename dof_data_t,
+             typename curl_layout_t, typename curl_data_t>
+   MFEM_ALWAYS_INLINE
+   void CalcCurl(const dof_layout_t  &dof_layout,
+                 const dof_data_t    &dof_data,
+                 const curl_layout_t &curl_layout,
+                 curl_data_t &curl_data) const
+   {
+      TMatrix<TDOF,1,real_t,true> dof_data_s;
+      TAssign<AssignOp::Set>(dof_layout, dof_data_s,dof_layout,dof_data);
+      AddOrientation(dof_data_s);
+      Calc<0,false>(dof_layout, dof_data_s,
+                    curl_layout.ind2(0), curl_data,
+                    B_1d_o.layout,  B_1d_o_neg,
+                    Gt_1d_c.layout, Gt_1d_c);
+
+      Calc<1,true>(dof_layout, dof_data_s,
+                   curl_layout.ind2(0), curl_data,
+                   G_1d_c.layout,  G_1d_c,
+                   Bt_1d_o.layout, Bt_1d_o);
+   }
+
+   // Multi-component gradient evaluation transpose from quadrature points to
+   // DOFs. grad_layout is (TNIP x DIM x NumComp), dof_layout is
+   // (TDOF x NumComp).
+   template <bool Add,
+             typename curl_layout_t, typename curl_data_t,
+             typename dof_layout_t, typename dof_data_t>
+   MFEM_ALWAYS_INLINE
+   void CalcCurlT(const curl_layout_t &curl_layout,
+                  const curl_data_t   &curl_data,
+                  const dof_layout_t  &dof_layout,
+                  dof_data_t          &dof_data) const
+   {
+      CalcT<0,Add>(curl_layout.ind2(0),curl_data,
+                   dof_layout, dof_data,
+                   Bt_1d_o.layout, Bt_1d_o_neg,
+                   G_1d_c.layout,  G_1d_c);
+      CalcT<1,Add>(curl_layout.ind2(0),curl_data,
+                   dof_layout, dof_data,
+                   Gt_1d_c.layout, Gt_1d_c,
+                   B_1d_o.layout,  B_1d_o);
+      AddOrientation(dof_data);
+   }
+
+   template <int D1, int D2, bool Add,
+             typename qpt_layout_t, typename qpt_data_t,
+             typename M_layout_t,  typename M_data_t,
+             typename D11_layout_t,typename D11_data_t,
+             typename D12_layout_t,typename D12_data_t,
+             typename D21_layout_t,typename D21_data_t,
+             typename D22_layout_t,typename D22_data_t>
+   MFEM_ALWAYS_INLINE
+   void AssembleBlock(const qpt_layout_t &qpt_layout,const qpt_data_t   &qpt_data,
```

Comment:
Would it be possible to have a few comments for this functions, even if I get the idea, it's really difficult to know what's really happening without any comment.